test/mgr: Improve unit-test coverage for ceph-mgr

stzuraski898 opened the pull request:
<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

The current test coverage and the unit-test coverage for ceph-mgr are not enough; changes that have been made sometimes can break the core functions of the mgr
We need to expand the test coverage for classes under src/mgr directory and expand the tests for the mgr modules.

Fixes: https://tracker.ceph.com/issues/72938

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

NitzanMordhai commented on src/test/mgr/CMakeLists.txt:
```diff
@@ -1,21 +1,112 @@
+  ${CMAKE_SOURCE_DIR}/src/mgr/ClusterState.cc
+)
+add_ceph_unittest(unittest_mgr_clusterstate)
+target_compile_definitions(unittest_mgr_clusterstate PRIVATE UNITTEST_BUILD)
```
@stzuraski898, please check UNIT_TESTS_BUILT, we have a few examples.

NitzanMordhai commented on src/test/mgr/CMakeLists.txt:
```diff
@@ -1,21 +1,112 @@
+  endif()
+  add_ceph_test(mgr-dashboard-smoke.sh
+    ${CMAKE_CURRENT_SOURCE_DIR}/mgr-dashboard-smoke.sh)
+endif()
```
Please explain, or revert if you didn't really want to change it.

stzuraski898 replied:
I was messing with my CMakeLists file to try and get around some errors and committed many things erroneously, I've removed this and some other changes that aren't needed

NitzanMordhai commented on src/test/mgr/test_clusterstate.cc:
```diff
@@ -0,0 +1,54 @@
+  ClusterState cs(mon_client, objecter, mgr_map);
+
+  SUCCEED();
+}
```
What will flag that test as failed?

NitzanMordhai commented on src/test/mgr/test_clusterstate.cc:
```diff
@@ -0,0 +1,54 @@
+  cs.set_mgr_map(mgr_map);
+  cs.set_service_map(service_map);
+
+  SUCCEED();
```
Same as above, what will flag it as failed?

NitzanMordhai commented on src/test/mgr/test_pyformatter.cc:
```diff
@@ -0,0 +1,210 @@
+  py_f.dump_int("item", 123);
+  py_f.close_section();
+  PyObject* py_obj = py_f.get();
+  ASSERT_NE(py_obj, nullptr);
```
Could you add a check here for the returned py_obj (using PyBytes_AsString)?

NitzanMordhai reviewed (changes requested):
@stzuraski898, please check my comments

stzuraski898 commented on src/test/mgr/test_daemonstate.cc:
```diff
@@ -0,0 +1,117 @@
+  device->set_metadata(std::move(metadata));
+
+  ASSERT_EQ(device->wear_level, 0.75f);
+  ASSERT_NE(device->life_expectancy.first, utime_t());
```
Will do!

NitzanMordhai reviewed (commented):
please check my comments

stzuraski898 commented on src/test/mgr/test_clusterstate.cc:
```diff
@@ -0,0 +1,54 @@
+  digest->mon_status_json.append("status");
+  cs.load_digest(digest.get());
+
+  SUCCEED();
```
Working on this and using TEST_F(TestMgr, ) to use the scaffolding's variables. Diagnosing a segfault that happens at the moment

stzuraski898 commented on src/test/mgr/TestMgr.h:
```diff
@@ -0,0 +1,91 @@
+
+    void SetUp() override
+    {
+      messenger = Messenger::create_client_messenger(g_ceph_context, "mgrtest");
```
Setup and teardown have been reworked n new commits, and I believe this has been resolved

NitzanMordhai commented on src/test/mgr/test_clusterstate.cc:
```diff
@@ -1,54 +1,325 @@
 // -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
 // vim: ts=8 sw=2 smarttab
 
+#include "common/ceph_argparse.h"
```
Do we really need it?

NitzanMordhai commented on src/test/mgr/test_clusterstate.cc:
```diff
@@ -1,54 +1,325 @@
+  ClusterState cs(mc, objecter, mgr_map);
+  
+  Objecter* new_objecter = new Objecter(g_ceph_context, messenger, mc, *icp);
+  cs.set_objecter(new_objecter);
```
TestMgr already has Objecter, no?

NitzanMordhai commented on src/test/mgr/test_clusterstate.cc:
```diff
@@ -1,54 +1,325 @@
+  s_map.epoch = 777;
+  cs.set_service_map(s_map);
+  cs.with_servicemap([&](const ServiceMap& sm){
+    ASSERT_EQ(sm.epoch, 777);
```
Use s_map.epoch in the compare

NitzanMordhai commented on src/test/mgr/test_clusterstate.cc:
```diff
@@ -1,54 +1,325 @@
-  SUCCEED();
+  cs.with_mon_status([&](const bufferlist& monstat){
+    ASSERT_EQ(monstat.length(), health_bl.length());
+    ASSERT_EQ(monstat.to_str(), "{\"monmap\":{\"epoch\":1}}");
```
use mon_status_bl for the compare

NitzanMordhai reviewed (changes requested):
@stzuraski898, please see my comments.  Can you also make sure that other comparisons use existing variables? There are a few more that i didnt commented.
lets also make sure that the we are using TestMgr to reduce the amount of repeated code, like objects that we use all over the tests.

stzuraski898 commented on src/test/mgr/test_mgrcap.cc:
```diff
@@ -299,3 +299,107 @@ TEST(MgrCap, Profile) {
+
+/* Begin Negative Tests */
+
+TEST(MgrCap, ParseEmptyString)
```
Which tests should be merged? I believe I merged all the negative tests that could be [here](https://github.com/ceph/ceph/pull/66432/commits/f814e4e77500d28aa13ad1dd70b4b84c4cd64fa2).

ParseGood, ParseBad, ParseIdentity all just check if the string parses.
The tests I haven't merged are making is_capable calls, which checks the description of an operation against capability. In the negative tests doing this, I'm sending descriptions that should fail, hence the ASSERT_FALSEs

NitzanMordhai reviewed (approved):
Looks much better!

ronen-fr commented on src/test/mgr/TestMgr.h:
```diff
@@ -0,0 +1,125 @@
+      mc = new MonClient(g_ceph_context, *icp);
+      objecter = new Objecter(g_ceph_context, messenger, mc, *icp);
+
+      assert(objecter != nullptr);
```
regular assert? you have used SetUp(), which means, I think, you can use ASSERT here

rzarzynski replied:
Conceptually it looks like a part of _the testing program's logic_, not necessarily of _the logic being tested_.

Anyway, `ceph_assert` is preferable.

ronen-fr commented on src/test/mgr/test_servicemap.cc:
```diff
@@ -0,0 +1,289 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/JSONFormatter.h"
```
same comment re order

ronen-fr commented on src/test/mgr/test_servicemap.cc:
```diff
@@ -0,0 +1,289 @@
+  std::string output = oss.str();
+
+  EXPECT_NE(output.find("\"gid\": 123"), std::string::npos);
+  EXPECT_NE(output.find("\"metadata\": {}"), std::string::npos);
```
We are compiling now with C++23, which means that all lines of the form
string.find() != npos
should be
string.contains()

ronen-fr reviewed (changes requested):
Basically - looks good.
I made some style comments, requesting changes that should be made in all the files (not just where I noted). These include:
- delay declaring variables to where they are needed. In many cases - that would allow you to mark them as const, too.
- clang-format (using the approved conf file ((or my conf file, which is better...)))
- use the RAII constructs of the Formatter(s) where appropiate.
- use 'contains' instead of fins!=npos
- ... a few more

rzarzynski commented on src/test/mgr/TestMgr.h:
```diff
@@ -0,0 +1,125 @@
+  protected:
+    ceph::ref_t<MPGStats> stats = ceph::make_ref<MPGStats>();
+    const mempool::osdmap::map<int64_t, pg_pool_t>& pools = osd_map.get_pools();
+    OSDMap::Incremental osd_inc{osd_map.get_epoch() + 1};
```
nit: let's be consistent and initialize with `=` or `{}` but without mixing the styles.

rzarzynski reviewed (commented):
Review-in-progress.

ronen-fr commented on src/test/mgr/test_daemonstate.cc:
```diff
@@ -0,0 +1,230 @@
+                    << "  expected: " << _e.sec() << "s " << _e.nsec() \
+                    << "ns\n";                                         \
+    }                                                                  \
+  } while (0);
```
minor nit: while (false)

ronen-fr commented on src/test/mgr/test_mgrcap.cc:
```diff
@@ -93,38 +91,37 @@ TEST(MgrCap, ParseGood) {
-  "allow command abc with arg=foo arg2=bar, allow service foo r",
-  0
-};
+const char* parse_identity[] = {
```
same comment (re char* to strings and loops) - here

ronen-fr reviewed (approved):
LGTM - apart of some 'C' style code.
See nits and comments.
A question: which .clang-format conf file did you use?

stzuraski898 commented on src/test/mgr/test_mgrcap.cc:
```diff
@@ -15,74 +15,72 @@
+TEST(MgrCap, ParseGood)
+{
+  for (int i = 0; parse_good[i]; ++i) {
     string str = parse_good[i];
```
These tests were existing already, but I agree and have updated them.

ronen-fr commented:
And - good work!

stzuraski898 commented on qa/valgrind.supp:
```diff
@@ -765,12 +765,70 @@
+  Memcheck:Leak
+  match-leak-kinds: all
+  ...
+  fun:UnknownInlinedFun
```
A function, specifically an inlined function that Valgrind cannot access the original function name of as a result of compiler optimizations from what I've read. In this case it is a CPython function being inlined. I may have generalized too far from the Valgrind flag --gen-suppressions=all, which was overly specific.

rzarzynski commented on src/test/mgr/test_clusterstate.cc:
```diff
@@ -0,0 +1,217 @@
+    bufferlist bl;
+    bl.append("{\"monmap\":{\"epoch\":1}}");
+    return bl;
+  }();
```
This reminds me `bufferlist` lacks a constructor for the cases like this one.
Anyway, nice workaround.

rzarzynski commented on src/test/mgr/test_mgrcap.cc:
```diff
@@ -15,164 +15,173 @@
-  "profile crash network 1.2.3.4/32",
-  "allow profile crash network 1.2.3.4/32",
-  0
+const std::vector<std::string> parse_good = {
```
I assume this only a **pure** refactor.

stzuraski898 committed:
test/mgr: Add TestMgr scaffolding to support unit tests

Fixes: https://tracker.ceph.com/issues/72938
Signed-off-by: szuraski898 <steven.zuraski@ibm.com>

stzuraski898 committed:
test/mgr: Add ClusterState unit tests

Fixes: https://tracker.ceph.com/issues/72938
Signed-off-by: szuraski898 <steven.zuraski@ibm.com>

stzuraski898 committed:
test/mgr: Add DaemonKey unit tests

Fixes: https://tracker.ceph.com/issues/72938
Signed-off-by: szuraski898 <steven.zuraski@ibm.com>

stzuraski898 committed:
test/mgr: Add DaemonState unit tests

Fixes: https://tracker.ceph.com/issues/72938
Signed-off-by: szuraski898 <steven.zuraski@ibm.com>

stzuraski898 committed:
test/mgr: Add PyFormatter unit tests

Fixes: https://tracker.ceph.com/issues/72938
Signed-off-by: szuraski898 <steven.zuraski@ibm.com>

stzuraski898 committed:
test/mgr: Add PyUtil unit tests

Fixes: https://tracker.ceph.com/issues/72938
Signed-off-by: szuraski898 <steven.zuraski@ibm.com>

stzuraski898 committed:
test/mgr: Add ServiceMap unit tests

Fixes: https://tracker.ceph.com/issues/72938
Signed-off-by: szuraski898 <steven.zuraski@ibm.com>

stzuraski898 committed:
test/mgr: Add negative unit tests to MgrCap

Fixes: https://tracker.ceph.com/issues/72938
Signed-off-by: szuraski898 <steven.zuraski@ibm.com>

stzuraski898 committed:
test/mgr: Add unit tests to CMake

Fixes: https://tracker.ceph.com/issues/72938
Signed-off-by: szuraski898 <steven.zuraski@ibm.com>

stzuraski898 committed:
test/mgr: clang-format changes to test_ttlcache

Fixes: https://tracker.ceph.com/issues/72938
Signed-off-by: szuraski898 <steven.zuraski@ibm.com>

stzuraski898 commented:
jenkins test make check arm64

stzuraski898 commented:
jenkins test api

stzuraski898 commented:
jenkins test make check arm64

SrinivasaBharath commented:
@stzuraski898-  Please resolve the pending check issues; once completed, you may proceed with merging the PR.

stzuraski898 commented:
jenkins test make check arm64

stzuraski898 commented:
jenkins test make check arm64

stzuraski898 commented:
jenkins test make check

stzuraski898 commented:
jenkins test make check

stzuraski898 commented:
jenkins test make check arm64

stzuraski898 committed:
qa: Valgrind suppression additions to pattern match inline functions stemming from CPython in mgr

Fixes: https://tracker.ceph.com/issues/72938
Signed-off-by: szuraski898 <steven.zuraski@ibm.com>

Conflicts:
  qa/valgrind.supp

The qa/valgrind.supp file had a conflict with an extra curly brace at the EOF as a result of changes adding CPython suppressions in the unittest-mgr branch, as well as additions coming form main.
Resolved by keeping both: the CPython suppressions for the mgr unit tests and all the rocky10 suppressions from main (rocky10_gcc14_mismatch_delete_map_erase
and rocky10 still reachable rocksdb leak), and removing the extra curly brace after merging.

stzuraski898 commented:
jenkins test make check arm64

stzuraski898 commented:
jenkins test make check arm64

stzuraski898 commented:
jenkins test make check arm64

stzuraski898 commented:
jenkins test make check

stzuraski898 commented:
jenkins test docs
